### PR TITLE
Add basic GameLogicPlugin

### PR DESCRIPTION
**Pull request by iche033**
Signed-off-by: Ian Chen <ichen@osrfoundation.org>

Added `GameLogicPlugin` with basic functionality for logging, starting and stopping a run, and publishing run status.

Added `test_game_logic` integration test.

Scoring functionality is not implemented yet.

There are 3 main plugin params:

* `<setup_duration_seconds>`: time allowed for teams to set up before starting a run
* `<run_duration_seconds>`: time limit for each run (count down starts when a run is "started")
* `<logging><path>`: Path to log files

**Starting and and finishing a run**

The actual procedure for starting  and stopping a run is still TBD. Currently service calls are provided to manually start and stop the run. To test:

```
ign service -s /mbzirc/start --reqtype ignition.msgs.Boolean --reptype ignition.msgs.Boolean --req 'data: true' --timeout 1000
```

```
ign service -s /mbzirc/finish --reqtype ignition.msgs.Boolean --reptype ignition.msgs.Boolean --req 'data: true' --timeout 1000
```

Invoking these service call should change the status of the run, which is published to the `/mbzirc/run_clock` topic:

```
ign topic -e -t /mbzirc/run_clock
```

Phases are: `setup` -> `run` -> `finished`.

**Logging**

4 logs are generated. They are stored in `/tmp/ign/mbzirc/logs` directory

* `events.yml`: Currently logs these events:
  * start
  * finish
  * dead battery
* `summary.yml`: Records whether the run was started, number of robots, total sim and real time of the run. This log file is updated periodically (30 seconds)
*  `score.yml`: scoring logic still need to be implemented so the log currently just outputs a score of `0`
* `mbzirc_<timestamp>.log`: general log msgs. Intended to be used for introspection, debugging, etc

### CODE REVIEW

File: mbzirc_ign/src/GameLogicPlugin.cc [hunk 1]
```diff
@@ -0,0 +1,653 @@
...
+  if (_sdf->HasElement("run_duration_seconds"))
+  {
+    this->dataPtr->runDuration = std::chrono::seconds(
+        _sdf->Get<int>("run_duration_seconds"));
+
+    ignmsg << "Run duration set to " << this->dataPtr->runDuration.count()
+           << " seconds.\n";
+  }
+
+  // Get the duration seconds.
```
Comment (caguero) [inline mbzirc_ign/src/GameLogicPlugin.cc:260]: setup duration?

File: mbzirc_ign/src/GameLogicPlugin.cc [hunk 2]
```diff
@@ -0,0 +1,653 @@
...
+      (common::joinPaths(this->dataPtr->logPath, filenamePrefix + "_" +
+      ignition::common::systemTimeISO() + ".log")).c_str(), std::ios::out);
+
+  // Open the event log file.
+  this->dataPtr->eventStream.open(
+      (common::joinPaths(this->dataPtr->logPath, "events.yml")).c_str(),
+      std::ios::out);
+
+  // Get the duration seconds.
+  if (_sdf->HasElement("run_duration_seconds"))
```
Comment (caguero) [inline mbzirc_ign/src/GameLogicPlugin.cc:251]: The default value of `runDuration` is 0. Should we increase that value or show an error message if the parameter isn't set?

File: mbzirc_ign/src/GameLogicPlugin.cc [hunk 3]
```diff
@@ -0,0 +1,653 @@
...
+  this->dataPtr->logStream.open(
+      (common::joinPaths(this->dataPtr->logPath, filenamePrefix + "_" +
+      ignition::common::systemTimeISO() + ".log")).c_str(), std::ios::out);
+
+  // Open the event log file.
+  this->dataPtr->eventStream.open(
+      (common::joinPaths(this->dataPtr->logPath, "events.yml")).c_str(),
+      std::ios::out);
+
+  // Get the duration seconds.
```
Comment (caguero) [inline mbzirc_ign/src/GameLogicPlugin.cc:250]: run duration?

File: mbzirc_ign/src/GameLogicPlugin.cc [hunk 4]
```diff
@@ -0,0 +1,653 @@
...
+  // Periodically update the score file.
+  if (!this->dataPtr->finished && currentTime -
+      this->dataPtr->lastUpdateScoresTime > std::chrono::seconds(30))
+  {
+    this->dataPtr->UpdateScoreFiles(this->dataPtr->simTime);
+  }
+}
+
+/////////////////////////////////////////////////
+void GameLogicPluginPrivate::PublishScore()
```
Comment (caguero) [inline mbzirc_ign/src/GameLogicPlugin.cc:459]: If the idea is to run this function in a separate thread I think we need a mutex to protect `totalScore`.

**Review by caguero — COMMENTED** [review]
It looks good to me, works as expected. Just a few minor comments.

File: mbzirc_ign/src/GameLogicPlugin.cc (hunk 1 above)
Comment (iche033) [reply mbzirc_ign/src/GameLogicPlugin.cc]: 969f88f5112dbd6d8c6894787bae404723c50a2c

File: mbzirc_ign/src/GameLogicPlugin.cc (hunk 2 above)
Comment (iche033) [reply mbzirc_ign/src/GameLogicPlugin.cc:260]: increased to 1 hr / 3600s. 969f88f5112dbd6d8c6894787bae404723c50a2c

File: mbzirc_ign/src/GameLogicPlugin.cc (hunk 3 above)
Comment (iche033) [reply mbzirc_ign/src/GameLogicPlugin.cc]: 969f88f5112dbd6d8c6894787bae404723c50a2c

File: mbzirc_ign/src/GameLogicPlugin.cc (hunk 4 above)
Comment (iche033) [reply mbzirc_ign/src/GameLogicPlugin.cc:472]: yeah added mutex to protect `totalScore` and also started a thread to publish score. 969f88f5112dbd6d8c6894787bae404723c50a2c

**Pull request merged**